### PR TITLE
Refactoring and general improvements

### PR DESCRIPTION
**Pull request by warrenbailey**
This PR contains the following:
1. Rename SecretStore to KeyStore and remove all secret functionality
2. Create a script that reads key content from a key folder and generates the secrets.xml
3. Refactor token_helper into specific JWT and JWE classes
4. Added test based on RFC example
5. Create custom Exception classes rather than using the base exception class
6. Updated documentation

Fixes the following issues #4 #9 and #10

### CODE REVIEW

File: sdc/crypto/jwe_helper.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                private_jwk = key_store.get_private_key_by_kid(purpose, jwe_kid).as_jwk()
+
+            jwe_token.decrypt(private_jwk)
+
+            return jwe_token.payload.decode()
+        except (ValueError, InvalidJWEData) as e:
+            raise InvalidTokenException(repr(e))
+
+    @staticmethod
+    def encrypt(payload, kid, key_store=None, purpose=None, alg="RSA-OAEP", enc="A256GCM", key=None):
```
Comment (collisdigital) [inline sdc/crypto/jwe_helper.py]: maybe `alg` and `enc` shouldn't be parameters if they should always be the default values for now (can always add them later if we want to support others?)

File: sdc/crypto/jwt_helper.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            signed_token.deserialize(jwt_token, key=public_jwk)
+
+            return json.loads(signed_token.claims)
+        except (InvalidJWSObject,
+                InvalidJWSSignature,
+                JWTInvalidClaimFormat,
+                JWTMissingClaim,
+                JWTExpired,
+                ValueError) as e:
+            raise InvalidTokenException(repr(e))
```
Comment (collisdigital) [inline sdc/crypto/jwt_helper.py]: chain exceptions with `raise InvalidTokenException from e`

File: sdc/crypto/scripts/generate_secrets.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        kid = hash_object.hexdigest()
+
+    key = {
+        "platform": platform,
+        "service": service,
+        "use": key_use,
+        "type": key_type,
+        "purpose": purpose,
+        "version": version,
+        "value": LiteralUnicode(private_key if private_key else public_key),
```
Comment (collisdigital) [inline sdc/crypto/scripts/generate_secrets.py]: maybe worth a sanity check/Exception that key_type matches this logic?

File: sdc/crypto/scripts/generate_secrets.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        except ValueError:
+            print("File {} is not in correct format".format(key_file))
+
+    generate_secrets_file(keys)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Generate secrets key file.')
+    parser.add_argument('folder', type=str,
```
Comment (collisdigital) [inline sdc/crypto/scripts/generate_secrets.py]: worth being able to set the output file path?

File: sdc/crypto/jwe_helper.py (hunk 1 above)
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: Agreed, I'll remove

File: sdc/crypto/scripts/generate_secrets.py (hunk 4 above)
Comment (warrenbailey) [inline sdc/crypto/scripts/generate_secrets.py]: Yes agreed, actually I'll add more output so it's easy to see which files have been processed and what file has been generated.

File: README.md [hunk 5]
```diff
@@ -21,38 +21,37 @@ Assuming you are executing from inside an activated virtual environment:
...
-# Load generated secret file and create secret store
-import yaml
-EXPECTED_SECRETS = []
-key_purpose = 'key_purpose'
+to store the file.  Then it needs to be loaded, and a key store generated.
+```
 
-with open('secrets.yml') as file:
-  secrets_from_file = yaml.safe_load(file)
+python sdc/crypto/scripts/generate_secrets.py <key_folder_location>
```
Comment (LJBabbage) [inline README.md]: This is the wrong location should be
```
sdc/crypto/utils/generate_secrets.py
```

File: sdc/crypto/jwt_helper.py (hunk 2 above)
Comment (JamesGardiner) [inline sdc/crypto/jwt_helper.py]: @warrenbailey - There are generally only two use cases where `from e` should be used in Python3:

1. Where error handling code results in another exception being raised
2. Where an exception gets deliberately raised as a different exception class (the case here)

Here's the PEP that introduced this https://www.python.org/dev/peps/pep-3134/

File: sdc/crypto/test/test_jwe_helper.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
+import os
```
Comment (ajmaddaford) [inline sdc/crypto/test/test_jwe_helper.py]: I think there are a number things we can do to tidy up these tests and make them clearer, but happy for them to be done in another pull request.

File: setup.py [hunk 7]
```diff
@@ -43,7 +43,7 @@
     ],
     packages=[
         "sdc.crypto",
-        "sdc.crypto.utils",
```
Comment (ajmaddaford) [inline setup.py]: Can we package the script up so it's available from the command line after installing the package? 
http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html

Comment (warrenbailey) [inline setup.py]: Done

File: sdc/crypto/test/test_jwe_helper.py (hunk 6 above)
Comment (warrenbailey) [inline sdc/crypto/test/test_jwe_helper.py]: Agreed, I did try but its quite a big job. I'll raise an issue specifically related to unit tests

File: sdc/crypto/exceptions.py [hunk 8]
```diff
@@ -1,4 +1,8 @@
 
+class CryptoError(Exception):
```
Comment (warrenbailey) [inline sdc/crypto/exceptions.py]: Done

File: sdc/crypto/jwe_helper.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+from jwcrypto import jwe
+from jwcrypto.jwe import InvalidJWEData
+from structlog import get_logger
```
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: Agreed, I'll remove

File: sdc/crypto/jwe_helper.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+logger = get_logger()
+
+
+class JWEHelper:
+    '''
+    Helper methods for encrypting a JWS/JWT. This class assumes you have used the RSA-OAP-A256GCM algoritm
+    and the required private key is available in the keystore.
+    '''
+    @staticmethod
+    def decrypt(encrypted_token, key_store=None, purpose=None, key=None):
```
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: Actually I had that originally but didn't like the decrypt_with_key method name. This is where I want function overloading.

I don't want to change the kwargs order as that's a api breaking change. So i'll add the new method as Andrew suggested.

File: sdc/crypto/test/test_jwe_helper.py (hunk 6 above)
Comment (warrenbailey) [inline sdc/crypto/test/test_jwe_helper.py]: https://github.com/ONSdigital/sdc-cryptography/issues/13

File: sdc/crypto/decrypter.py [hunk 11]
```diff
@@ -1,13 +1,14 @@
...
-def decrypt(token, secret_store, key_purpose, leeway=120):
+def decrypt(token, key_store, key_purpose, leeway=120):
     """This decrypts the provided jwe token and the resulting jwt token and returns
     the payload.
 
     :param str token: The jwe token.
-    :param secret_store: The secret store.
+    :param key_store: The secret store.
     :param str key_purpose: Context for the key.
     :param int leeway: Extra allowed time after expiration to account for clock skew.
```
Comment (warrenbailey) [inline sdc/crypto/decrypter.py]: seconds

File: sdc/crypto/decrypter.py [hunk 12]
```diff
@@ -1,13 +1,14 @@
...
 
 
-def decrypt(token, secret_store, key_purpose, leeway=120):
+def decrypt(token, key_store, key_purpose, leeway=120):
     """This decrypts the provided jwe token and the resulting jwt token and returns
     the payload.
 
     :param str token: The jwe token.
-    :param secret_store: The secret store.
+    :param key_store: The secret store.
```
Comment (warrenbailey) [inline sdc/crypto/decrypter.py]: done

File: sdc/crypto/decrypter.py [hunk 13]
```diff
@@ -1,13 +1,14 @@
...
+from sdc.crypto.jwe_helper import JWEHelper
+from sdc.crypto.jwt_helper import JWTHelper
 
 
-def decrypt(token, secret_store, key_purpose, leeway=120):
+def decrypt(token, key_store, key_purpose, leeway=120):
     """This decrypts the provided jwe token and the resulting jwt token and returns
     the payload.
 
     :param str token: The jwe token.
```
Comment (warrenbailey) [inline sdc/crypto/decrypter.py]: Done

File: sdc/crypto/jwe_helper.py (hunk 10 above)
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: Done

File: sdc/crypto/jwe_helper.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        try:
+            jwe_token = jwe.JWE(algs=['RSA-OAEP', 'A256GCM'])
+            jwe_token.deserialize(encrypted_token)
+
+            if key:
+                logger.info("Decrypting JWE with provided key")
+                private_jwk = key
+            else:
+                jwe_kid = extract_kid_from_header(encrypted_token)
```
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: Done

File: sdc/crypto/jwe_helper.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+                private_jwk = key_store.get_private_key_by_kid(purpose, jwe_kid).as_jwk()
+
+            jwe_token.decrypt(private_jwk)
+
+            return jwe_token.payload.decode()
+        except (ValueError, InvalidJWEData) as e:
+            raise InvalidTokenException(str(e)) from e
+
+    @staticmethod
+    def encrypt(payload, kid, key_store=None, purpose=None, key=None):
```
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: Adding an encrypt_with_key method

File: sdc/crypto/keys.py [hunk 16]
```diff
@@ -48,18 +40,14 @@ def as_jwk(self):
         return jwk.JWK.from_pem(self.value.encode('utf-8'))
 
 
-class SecretStore:
+class KeyStore:
```
Comment (warrenbailey) [inline sdc/crypto/keys.py]: I'm going to leave this one, I don't think it's needed

File: sdc/crypto/encrypter.py [hunk 17]
```diff
@@ -1,19 +1,20 @@
...
+from sdc.crypto.jwt_helper import JWTHelper
 
 
-def encrypt(json, secret_store, key_purpose):
+def encrypt(json, key_store, key_purpose):
     """This encrypts the supplied json and returns a jwe token.
 
     :param str json: The json to be encrypted.
-    :param secret_store: The secret store.
+    :param key_store: The secret store.
```
Comment (warrenbailey) [inline sdc/crypto/encrypter.py]: Done

File: sdc/crypto/decrypter.py (hunk 13 above)
Comment (warrenbailey) [inline sdc/crypto/decrypter.py]: Done

File: sdc/crypto/decrypter.py (hunk 11 above)
Comment (warrenbailey) [inline sdc/crypto/decrypter.py]: Seconds - updated docs

File: sdc/crypto/keys.py [hunk 18]
```diff
@@ -48,18 +40,14 @@ def as_jwk(self):
...
     def __init__(self, secrets):
```
Comment (warrenbailey) [inline sdc/crypto/keys.py]: Done

File: sdc/crypto/scripts/generate_secrets.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+
+    pub_bytes = pub_key.public_bytes(Encoding.PEM, PublicFormat.SubjectPublicKeyInfo)
+
+    _generate_kid_from_key_and_add_to_dict(keys, "private", platform, service, purpose, key_use, version,
+                                           pub_bytes.decode(), private_key_data, kid_override=kid_override)
+
+
+def generate_secrets_file(keys, ):
+    with open('secrets.yml', 'w') as f:
+        yaml.dump({"keys": keys}, f, default_flow_style=False)
```
Comment (warrenbailey) [inline sdc/crypto/scripts/generate_secrets.py]: I can't find the documentation you are referring to?

File: sdc/crypto/scripts/generate_secrets.py [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+
+    pub_key = private_key.public_key()
+
+    pub_bytes = pub_key.public_bytes(Encoding.PEM, PublicFormat.SubjectPublicKeyInfo)
+
+    _generate_kid_from_key_and_add_to_dict(keys, "private", platform, service, purpose, key_use, version,
+                                           pub_bytes.decode(), private_key_data, kid_override=kid_override)
+
+
+def generate_secrets_file(keys, ):
```
Comment (warrenbailey) [inline sdc/crypto/scripts/generate_secrets.py]: Done

File: sdc/crypto/scripts/generate_secrets.py [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+    keys = {}
+
+    for key_file in key_files:
+        try:
+            print("Processing key file {}".format(key_file))
+            # remove the .pem extension
+            key_file_no_pem = key_file[:key_file.index(".")]
+            platform, service, purpose, key_use, key_type, version = key_file_no_pem.split("-")
+            if key_type == "public":
+                add_public_key_to_dict(keys, platform, service, purpose, key_use, version, key_file, keys_folder)
```
Comment (warrenbailey) [inline sdc/crypto/scripts/generate_secrets.py]: Agreed, I didn't like this either.

**Comment by warrenbailey** [comment]
@ajmaddaford ready for re-review

File: sdc/crypto/jwe_helper.py [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+from sdc.crypto.exceptions import InvalidTokenException, MissingKidException, MissingKeyException
+from sdc.crypto.helper import extract_kid_from_header
+
+logger = logging.getLogger(__name__)
+
+
+class JWEHelper:
+    '''
+    Helper methods for encrypting a JWS/JWT. This class assumes you have used the RSA-OAP-A256GCM algoritm
```
Comment (JamesGardiner) [inline sdc/crypto/jwe_helper.py]: Typo in algoritm

File: sdc/crypto/scripts/generate_keys.py [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    :param public_key: The name of the public key to add
+    :param keys_folder: The location on disk where the key exists
+    :param kid_override: This allows the caller to override the generated KID value
+    :return: None
+    '''
+    public_key_data = get_file_contents(keys_folder, public_key)
+
+    kid = _generate_kid_from_key(public_key_data)
+
+    key = _create_key(platform=platform, service=service, key_use=key_use, key_type="public", purpose=purpose,
```
Comment (JamesGardiner) [inline sdc/crypto/scripts/generate_keys.py]: Line too long?

File: sdc/crypto/scripts/generate_keys.py [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+
+    private_key = load_pem_private_key(private_key_data.encode(), None, backend=backend)
+
+    pub_key = private_key.public_key()
+
+    pub_bytes = pub_key.public_bytes(Encoding.PEM, PublicFormat.SubjectPublicKeyInfo)
+
+    kid = _generate_kid_from_key(pub_bytes.decode())
+
+    key = _create_key(platform=platform, service=service, key_use=key_use, key_type="private", purpose=purpose,
```
Comment (JamesGardiner) [inline sdc/crypto/scripts/generate_keys.py]: Break line?

File: sdc/crypto/exceptions.py [hunk 25]
```diff
@@ -0,0 +1,23 @@
+
+class CryptoError(Exception):
+    ''' General Exception raise by this library if an error occurs not covered below '''
+    pass
+
+
+class MissingKidException(Exception):
+    ''' Raised if a Kid cannot be found in the keystore '''
```
Comment (necrophonic) [inline sdc/crypto/exceptions.py]: Isn't this raised if the Kid is not found in the message header rather than in the keystore?

Comment (warrenbailey) [inline sdc/crypto/exceptions.py]: Updated doc string

File: README.md [hunk 26]
```diff
@@ -20,39 +20,38 @@ Assuming you are executing from inside an activated virtual environment:
...
 
-with open('secrets.yml') as file:
-  secrets_from_file = yaml.safe_load(file)
+generate_keys.py <key_folder_location>
 
-validate_required_secrets(secrets_from_file, EXPECTED_SECRETS, key_purpose)
-secret_store = SecretStore(secrets_from_file)
 ```
 
 After this has been configured, encrypting and decrypting can be done like in the
```
Comment (JamesGardiner) [inline README.md]: as in the example below

File: README.md [hunk 27]
```diff
@@ -20,39 +20,38 @@ Assuming you are executing from inside an activated virtual environment:
...
 example below.
 ```python
+
+secrets_from_file = yaml.safe_load("secrets.yml")
+
+validate_required_secrets(secrets_from_file, EXPECTED_SECRETS, KEY_PURPOSE_SUBMISSION)
+
+key_store= KeyStore(secrets_from_file)
```
Comment (JamesGardiner) [inline README.md]: space after var name

File: README.md [hunk 28]
```diff
@@ -20,39 +20,38 @@ Assuming you are executing from inside an activated virtual environment:
...
 from sdc.crypto.decrypter import decrypt
 data_bytes = data.decode('UTF8')
-decrypted_json = decrypt(data_bytes, secret_store, key_purpose)
+decrypted_json = decrypt(data_bytes, key_store, key_purpose)
 
 ```
+
+###### PyPi
+
+This repo is available in PyPi at [sdc-cryptography]https://pypi.python.org/pypi/sdc-cryptography)
```
Comment (JamesGardiner) [inline README.md]: from pypi

File: sdc/crypto/exceptions.py [hunk 29]
```diff
@@ -0,0 +1,23 @@
+
+class CryptoError(Exception):
+    ''' General Exception raise by this library if an error occurs not covered below '''
```
Comment (JamesGardiner) [inline sdc/crypto/exceptions.py]: raised

File: sdc/crypto/helper.py [hunk 30]
```diff
@@ -0,0 +1,23 @@
+import json
+
+from jwcrypto.common import base64url_decode
+from sdc.crypto.exceptions import InvalidTokenException
+
+
+def extract_kid_from_header(token):
+    header = token.split('.')[:1][0]
+
+    if header is "":
```
Comment (JamesGardiner) [inline sdc/crypto/helper.py]: `if not header`

File: sdc/crypto/jwe_helper.py [hunk 31]
```diff
@@ -0,0 +1,93 @@
+import logging
+from jwcrypto import jwe
```
Comment (JamesGardiner) [inline sdc/crypto/jwe_helper.py]: space between stdlib imports and external packages

File: sdc/crypto/jwe_helper.py [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+from jwcrypto.jwe import InvalidJWEData
+
+
+from sdc.crypto.exceptions import InvalidTokenException, MissingKidException, MissingKeyException
```
Comment (JamesGardiner) [inline sdc/crypto/jwe_helper.py]: InvalidTokenException, MissingKeyException,  MissingKidException

File: sdc/crypto/jwe_helper.py [hunk 33]
```diff
@@ -0,0 +1,93 @@
...
+    and the required private key is available in the keystore.
+    '''
+    @staticmethod
+    def decrypt(encrypted_token, key_store=None, purpose=None):
+        try:
+            jwe_token = jwe.JWE(algs=['RSA-OAEP', 'A256GCM'])
+            jwe_token.deserialize(encrypted_token)
+
+            jwe_kid = extract_kid_from_header(encrypted_token)
+            if not jwe_kid:
```
Comment (JamesGardiner) [inline sdc/crypto/jwe_helper.py]: Does this function not raise an error if kid is None?

Comment (JamesGardiner) [inline sdc/crypto/jwe_helper.py]: https://github.com/ONSdigital/sdc-cryptography/pull/12/files#diff-1d5ce69bad7559f75665ca2226128f93R23

File: sdc/crypto/jwe_helper.py [hunk 34]
```diff
@@ -0,0 +1,93 @@
...
+                raise MissingKidException
+
+            logger.info("Decrypting JWE kid is {}".format(jwe_kid))
+
+            private_jwk = key_store.get_private_key_by_kid(purpose, jwe_kid).as_jwk()
+            if not private_jwk:
```
Comment (JamesGardiner) [inline sdc/crypto/jwe_helper.py]: Same point as above - do we need to check for None?

File: setup.py (hunk 7 above)
Comment (jonnyshaw89) [inline setup.py]: The script works fine with our existing data.

File: sdc/crypto/jwe_helper.py [hunk 35]
```diff
@@ -0,0 +1,93 @@
...
+                raise MissingKeyException
+
+            jwe_token.decrypt(private_jwk)
```
Comment (jonnyshaw89) [inline sdc/crypto/jwe_helper.py]: If we call `decrypt_with_key` here then we can get rid of some duplication at the top of this method.

File: sdc/crypto/helper.py (hunk 30 above)
Comment (warrenbailey) [inline sdc/crypto/helper.py]: Good spot

File: sdc/crypto/jwe_helper.py (hunk 33 above)
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: You've lost me here the next line::
raise MissingKidException??

File: sdc/crypto/jwe_helper.py (hunk 35 above)
Comment (warrenbailey) [inline sdc/crypto/jwe_helper.py]: I did try that, but it fails loads of tests because the tests assume that the text in the exception matches that generated by the underlying library in the call to   jwe_token.deserialize(encrypted_token)

It's a lot of effort to change, but agree with you it would be better that way. I'll add it to the test issue here: https://github.com/ONSdigital/sdc-cryptography/issues/13 and fix it then
